File: android/src/main/java/com/tolotra/account_picker/AccountPickerPlugin.java
package com.tolotra.account_picker;

import android.app.Activity;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** AccountPickerPlugin */
public class AccountPickerPlugin implements FlutterPlugin, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;
  private ActivityPluginBinding activityBinding;
  private Activity activity;
  private FlutterPluginBinding pluginBinding;

  @Nullable
  private MethodCallHandlerImpl methodCallHandler;
  private AccountPicker location;


  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding binding) {
    pluginBinding = binding;

    location = new AccountPicker(binding.getApplicationContext(), /* activity= */ null);
    methodCallHandler = new MethodCallHandlerImpl(location);
    methodCallHandler.startListening(binding.getBinaryMessenger());
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    AccountPicker flutterLocation = new AccountPicker(registrar);
    flutterLocation.setActivity(registrar.activity());

    MethodCallHandlerImpl handler = new MethodCallHandlerImpl(flutterLocation);
    handler.startListening(registrar.messenger());
  }


  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {

    pluginBinding = null;

    if (methodCallHandler != null) {
      methodCallHandler.stopListening();
      methodCallHandler = null;
    }

    location = null;
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    location.setActivity(binding.getActivity());
    activityBinding = binding;
    setup(pluginBinding.getBinaryMessenger(), activityBinding.getActivity(), null);
  }

  @Override
  public void onDetachedFromActivity() {
    tearDown();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {
    onDetachedFromActivity();
  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {
    onAttachedToActivity(binding);
  }

  private void setup(final BinaryMessenger messenger, final Activity activity,
                     final PluginRegistry.Registrar registrar) {
    if (registrar != null) {
      // V1 embedding setup for activity listeners.
      registrar.addActivityResultListener(location);
    } else {
      // V2 embedding setup for activity listeners.
      activityBinding.addActivityResultListener(location);
    }
  }

  private void tearDown() {
    activityBinding.removeActivityResultListener(location);
  }
}

File: android/src/main/kotlin/com/tolotra/account_picker/AccountPickerPlugin.kt
package com.tolotra.account_picker

import android.app.Activity
import android.content.Context
import android.os.Build
import androidx.activity.ComponentActivity
import androidx.annotation.RequiresApi
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** AccountPickerPlugin */
class AccountPickerPlugin : AccountPickerPluginMethod(), ActivityAware {


    // ActivityAware Methods
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        val componentActivity = activity as? ComponentActivity
            ?: throw IllegalStateException("Activity must extend ComponentActivity to use registerForActivityResult")

        accountPicker = AccountPicker(componentActivity)
//        binding.addActivityResultListener(accountPicker)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        accountPicker = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity

        val componentActivity = activity as? ComponentActivity
            ?: throw IllegalStateException("Activity must extend ComponentActivity to use registerForActivityResult")

        accountPicker = AccountPicker(componentActivity)
//        binding.addActivityResultListener(accountPicker!!)
    }

    override fun onDetachedFromActivity() {
        activity = null
        accountPicker = null
    }
}

abstract class AccountPickerPluginMethod : AccountPickerPluginEngine() {
    @RequiresApi(Build.VERSION_CODES.M)
    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getPlatformVersion") {
            result.success("Android ${android.os.Build.VERSION.RELEASE}")
        } else if (call.method == "requestEmailHint") {
            accountPicker?.requestEmailHint(result)
        } else if (call.method == "requestPhoneHint") {
            accountPicker?.requestPhoneHint(result)
        } else {
            result.notImplemented()
        }
    }
}

abstract class AccountPickerPluginEngine : FlutterPlugin, MethodCallHandler {

    /// Reference to the Activity and Context
    var activity: Activity? = null
    var context: Context? = null
    var accountPicker: AccountPicker? = null

    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "account_picker")
        channel.setMethodCallHandler(this)
    }


    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

}
